Update docs for "add_additional_resource" and other fixes

GraemeWatt opened the pull request:
I finally got around to addressing #197 that I opened in June 2022.  In doing so, I encountered many basic problems with my local installation (on macOS 13.5.2 with Python 3.11) including running the tests and building the docs.  I've tried to fix all the problems that I encountered.  Consequently, the scope of this PR is much larger than originally intended.  I'll try to summarise the problems and my fixes below.

* I've added new sections "Adding resource links or files" in `usage.rst` to explain the `add_additional_resource` method of the `Submission` and `Table` objects.  I extended the method to support an additional argument `file_type="HistFactory"` and modified the `test_additional_resource_size` function.  I modified all five example Jupyter notebooks to use the `add_additional_resource` method.  Closes #197.
* I added ROOT v6.28 to the test matrix in `tests.yml`.
* I changed https://zenodo.org/record/4946277 (v0.6.0 of `hepdata_lib`) written in the `__init__` method of the `Submission` class to https://doi.org/10.5281/zenodo.1217998 (the Concept DOI).  Closes #212.
* I've updated `.readthedocs.yml` following current recommendations and based on experience with the main [HEPData/hepdata](https://github.com/HEPData/hepdata) repo (closes #227).  In particular, installing `hepdata_lib` should (hopefully!) build the module index (closes #138).  I won't know if it works until I open this PR, and some further tweaks might be needed, so I'm going to open a draft PR to start with.
* I added the `hepdata_lib.c_file_reader` and `hepdata_lib.helpers` modules to the docs.
* The method for running the tests `python setup.py test` mentioned in `dev.rst` is deprecated and it does not even work for me ("Ran 0 tests in 0.000s").  It is not used in the CI (`tests.yml`) which instead uses `python -m pytest tests`.  I've changed `dev.rst` to instead recommend using `pytest tests`.  I didn't change another instance of `python setup.py test` in `release.yml` which does not seem to be used recently.
* The CI (`tests.yml`) has a command [`python -m pip install '.[test]'`](https://github.com/HEPData/hepdata_lib/blob/9330eb9e810dde04ce45504bc89be98e1f2e05c9/.github/workflows/tests.yml#L79-L80) which gives `WARNING: hepdata-lib 0.13.0 does not provide the extra 'test'`.  This is because `extras_require` needs to be defined in `setup.py`, which I've now done with the requirements needed to run the tests locally on my system.  The existing CI still works despite this bug because the test requirements are all installed in the [previous step](https://github.com/HEPData/hepdata_lib/blob/9330eb9e810dde04ce45504bc89be98e1f2e05c9/.github/workflows/tests.yml#L63-L77).  I've also removed the deprecated `setup_requires` and `tests_require`.
* I've added a new section "Building the documentation" at the end of `dev.rst`.  I added `Sphinx<7` to the `docs/requirements.txt` file as the build fails with Sphinx v7.  I replaced `m2r2` with `sphinx-mdinclude` to avoid a version conflict with Mistune when installing Jupyter in the same virtual environment.
* Running Sphinx locally gave many warnings/errors, for example, due to missing line breaks.  I've added `SPHINXOPTS    = -W --keep-going` to the Makefile and `fail_on_warning: true` to `.readthedocs.yml`.  This should help avoid building broken docs.  For example, the links to the Jupyter notebooks and Binder links in [Further examples](https://hepdata-lib.readthedocs.io/en/latest/README.html#further-examples) are not working properly.
* I've added an explanation to `setup.rst` to mention use of the standard `venv` module rather than `virtualenv`/`virtualenvwrapper`.
* I managed to get almost everything working under Python v3.11, so I extended the test matrix and pushed to GitHub.  Then I realised I should run Pylint and found that the installed v2.9.6 is not compatible with Python 3.11, because it installs `wrapt` v1.12.1 via `astroid`.  We could upgrade Pylint to a more recent version, but the `pylintrc` files would need to be updated, since they use many obsolete options.  For now, I removed support for Python v3.11 and kept Pylint v2.9.6.  I added a section on "Analysing the code" in `dev.rst` to get the developer to run the same commands as used in GitHub Actions.  I got several error messages that I didn't understand, so for now I simply added `# pylint: disable-msg` comments to the offending lines.

<!-- readthedocs-preview hepdata-lib start -->
----
:books: Documentation preview :books:: https://hepdata-lib--233.org.readthedocs.build/en/233/

<!-- readthedocs-preview hepdata-lib end -->

matthewfeickert commented:
@GraemeWatt I'm about to go on holiday for a week and I won't have internet access, so while I might be able to do a first pass on this before I go to bed tonight I think I'll defer everything to your judgement and @clelange review. Thanks for this PR!

GraemeWatt commented:
> @GraemeWatt I'm about to go on holiday for a week and I won't have internet access, so while I might be able to do a first pass on this before I go to bed tonight I think I'll defer everything to your judgement and @clelange review. Thanks for this PR!

Sorry, I meant to comment that your review would be optional and only if you had spare time.  I just thought I'd add you as a Reviewer since you contributed to improving this repository earlier in the year.

matthewfeickert commented on setup.py:
```diff
@@ -37,14 +37,17 @@
     packages=find_packages(exclude=['contrib', 'docs', 'tests']),
     zip_safe=False,
-    python_requires='>=3.6',
+    python_requires='>=3.6,<3.11',
```
Capping Python requires should be avoided for reasons given by Henry here: https://iscinumpy.dev/post/bound-version-constraints/

GraemeWatt replied:
Thanks for the feedback and your dedication to prompt PR reviewing!  Enjoy your vacation!

I agree to remove the cap `<3.11`.  My reasoning was that installation should not be allowed for Python versions not included in the test matrix.  However, in this case, I checked locally that the tests pass with Python 3.11.  It's only Pylint that causes problems with Python 3.11, so I've opened a new issue #234.

matthewfeickert commented on docs/setup.rst:
```diff
@@ -47,7 +47,7 @@ If you would like to develop the code, you need to install the package from the
 
     cd $SOMEPATH
-    git clone git@github.com:HEPData/hepdata_lib.git
+    git clone https://github.com/HEPData/hepdata_lib.git
```
Switching to HTTPS is fine, though I like poking people towards using SSH when possible.

GraemeWatt replied:
I guess it's personal preference whether to use HTTPS or SSH, so I've given both URLs and linked to the GitHub documentation.  I just changed to HTTPS because that's what I use myself.

clelange replied:
I'd also prefer to keep SSH here

matthewfeickert reviewed (changes requested):
Thanks @GraemeWatt! This all looks good with the exception of the cap on requires-python metadata. This review comes from plane WiFi and the GitHub mobile app so I apologize for not giving better inline suggestions or clarity.

clelange commented on docs/setup.rst:
```diff
@@ -42,12 +42,18 @@ Setup for developers
     cd $SOMEPATH
-    git clone git@github.com:HEPData/hepdata_lib.git
+    git clone https://github.com/HEPData/hepdata_lib.git
+ OR git clone git@github.com:HEPData/hepdata_lib.git
```
... and then make the HTTPS option a comment so one can copy and paste

GraemeWatt replied:
OK, I've reverted to giving SSH, with the HTTPS option given in the text above.  Personally, I just find to easier to use HTTPS since it doesn't require an SSH keypair, especially if I just want to checkout a repository without pushing to it.

clelange commented on docs/usage.rst:
```diff
@@ -236,12 +264,25 @@ If you have the full-size plot available on your drive, you can add it to your e
 
 The library code then takes care of all the necessary steps, like converting the image to the right format and size, and copying it into your submission folder. The conversion relies on the ImageMagick library, and will only work if the ``convert`` command is available on your machine.
 
+Adding resource links or files
```
This now seems a repetition of what is above, but with less detail? Would suggest removing it here.

GraemeWatt replied:
The repetition is deliberate because the `add_additional_resource` method can be called from an instance of either the `Submission` or `Table` class, and people might only look at the relevant section.  I restructured the first sentence to move "In the same way as for the Submission object" to the start.

GraemeWatt commented on README.md:
```diff
@@ -53,11 +57,25 @@ Unpacking the image can take a few minutes the first time you use it. Please be
-- [Reading in ROOT histograms](examples/reading_histograms.ipynb) [![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/HEPData/hepdata_lib/master?filepath=examples/reading_histograms.ipynb)
-- [Reading a correlation matrix](examples/correlation.ipynb) [![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/HEPData/hepdata_lib/master?filepath=examples/correlation.ipynb)
-- [Reading TGraph and TGraphError from '.C' files](examples/read_c_file.ipynb) [![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/HEPData/hepdata_lib/master?filepath=examples/read_c_file.ipynb)
+- [Reading in text files](https://github.com/HEPData/hepdata_lib/blob/master/examples/Getting_started.ipynb)\
```
The problem is that the `README.md` file is rendered in two different places.  Firstly, on the main page of the repo (https://github.com/HEPData/hepdata_lib#further-examples) where it displays fine.  But the `README.md` file is also converted to RST format for rendering on *Read the Docs* (https://hepdata-lib.readthedocs.io/en/latest/README.html#further-examples).  This is currently broken.  The relative links do not work giving "404 Not Found" from *Read the Docs*.  The Binder images/links also do not render properly.  The [Read the Docs build](https://readthedocs.org/api/v2/build/21721306.txt) gives error/warning messages like:
```
../README.md:96: ERROR: Unexpected indentation.
../README.md:101: ERROR: Unexpected indentation.
../README.md:106: ERROR: Unexpected indentation.
../README.md:111: ERROR: Unexpected indentation.
../README.md:116: ERROR: Unexpected indentation.
/home/docs/checkouts/readthedocs.org/user_builds/hepdata-lib/checkouts/stable/docs/contributing.rst:75: WARNING: Title underline too short.
```
These messages are also apparent when running Sphinx locally.  My fixes are a compromise to get rid of these error messages and ensure an acceptable rendering on both GitHub and *Read the Docs*.  Admittedly, the spacing now looks worse than before on GitHub, but it was the best I could do.  Now I've enabled `fail_on_warning: true` in the *Read the Docs* configuration, such error messages will cause the build to fail rather than being hidden away.

GraemeWatt replied:
I tried to improve the layout of the "Further examples" section in my latest [commit](https://github.com/HEPData/hepdata_lib/commit/497d462f2639f1b96ae1d35e0432cdbe824e1726).  Now the Binder images/links appear on the same line as the text on [GitHub](https://github.com/HEPData/hepdata_lib/tree/update-docs#further-examples) (as before).  On [*Read the Docs*](https://hepdata-lib--233.org.readthedocs.build/en/233/README.html#further-examples), the Binder images/links appear on the line below the text.  The two line breaks `<br/><br/>` are needed to increase the separation after each bullet point because otherwise the Binder images/links appear immediately above the text of the next bullet point.

GraemeWatt commented on docs/setup.rst:
```diff
@@ -92,6 +98,15 @@ You can always activate the virtual environment in another shell by calling the
     workon hepdata_git
     python myscript.py # Execute script using development branch
 
+Alternatively, if you don't want to install the additional virtualenv_ and virtualenvwrapper_ packages, you can simply
```
Agree, I think the text on `virtualenv`/`virtualenvwrapper` is a hangover from Python 2 where the `venv` module was not available.  I've rewritten the text to use the `venv` module instead.

GraemeWatt commented:
> Hi @GraemeWatt - thanks, this changes quite a lot, but I agree with @matthewfeickert that things overall look good. I left a couple of suggestions/comments in the code.
> 
> Two general questions:
> 
> * Why did you change the formatting in so many places, in particular those lines/separators?
> * Do we really need to disable pylint for missing imports in so many places (`# pylint: disable-msg=E0401`) or is this a setup issue on your side?

Thanks for the useful review!  I've commented separately on your individual points.  The formatting changes were necessary to avoid errors/warnings about incorrectly formatted `.rst` files when running Sphinx.  I've now removed the `# pylint: disable-msg` comments that I added previously and fixed the Pylint errors by making some other changes (adding `__init__.py`, importing `test_utilities.py` as a relative module after `hepdata_lib`, adding `math` and `array` modules to `extension-pkg-whitelist` in `pylintrc`).  My setup is not unusual (macOS 13.5.2 with Python 3.10), so I don't know why these errors didn't show up previously in the CI.

I also renamed the `master` branch to `main` as for other HEPData repositories (hope you don't mind).  I updated the actions in the GitHub Actions workflows to hopefully close some of the [open Dependabot PRs](https://github.com/HEPData/hepdata_lib/pulls/app%2Fdependabot).

Is the `release.yml` workflow currently being used?  If not, removing it from the repository would reduce maintenance.  It seems to use many unofficial actions.  I didn't update the `python setup.py test` line.  If there's a reason why actions should not be upgraded, the `dependabot.yml` file should have an [`ignore`](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#ignore) option.

clelange commented:
>  Is the `release.yml` workflow currently being used? If not, removing it from the repository would reduce maintenance. It seems to use many unofficial actions. I didn't update the `python setup.py` test line. If there's a reason why actions should not be upgraded, the dependabot.yml file should have an [`ignore`](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#ignore) option.

Yes, we should remove this action, there are better ones available now. I'll create an issue.

clelange commented:
Actually, we have this in https://github.com/HEPData/hepdata_lib/issues/165 already, I renamed it slightly.

clelange commented:
Thanks for addressing my comments, I'll merge this now.
